ccl/kvccl/kvfollowerreadsccl: TestSecondaryTenantFollowerReadsRouting failed

ccl/kvccl/kvfollowerreadsccl.TestSecondaryTenantFollowerReadsRouting [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8335731?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8335731?buildTab=artifacts#/) on master @ [43cd78ecd87591d9b1d3b142f29ddfd76013b276](https://github.com/cockroachdb/cockroach/commits/43cd78ecd87591d9b1d3b142f29ddfd76013b276):


```
        	            	     8.898ms      0.674ms    event:sql/conn_executor_exec.go:2268 [nsql4,client=127.0.0.1:44690,user=root] AutoCommit. err: <nil>
I230117 06:23:01.204863 1215768 testutils/testcluster/testcluster.go:146  [T1] 1835  TestCluster quiescing nodes
I230117 06:23:01.205596 1382849 sql/stats/automatic_stats.go:508  [T1,n2] 1836  quiescing auto stats refresher
I230117 06:23:01.206863 1425954 sql/stats/automatic_stats.go:508  [T1,n3] 1843  quiescing auto stats refresher
W230117 06:23:01.207414 1744764 sql/sqlliveness/slinstance/slinstance.go:312  [T1,nsql4] 1845  exiting heartbeat loop
I230117 06:23:01.206045 1464911 sql/stats/automatic_stats.go:508  [T1,n4] 1838  quiescing auto stats refresher
W230117 06:23:01.206336 1465418 jobs/registry.go:1007  [T1,n4] 1839  canceling all adopted jobs due to stopper quiescing
W230117 06:23:01.205673 1425482 sql/sqlliveness/slinstance/slinstance.go:312  [T1,n3] 1840  exiting heartbeat loop
W230117 06:23:01.206767 1745694 jobs/registry.go:1007  [T10,nsql4] 1841  canceling all adopted jobs due to stopper quiescing
I230117 06:23:01.206650 1745655 sql/stats/automatic_stats.go:508  [T10,nsql4] 1842  quiescing auto stats refresher
W230117 06:23:01.209653 1664400 ccl/kvccl/kvtenantccl/connector.go:235  [T10,nsql2,tenant-connector] 1844  error consuming GossipSubscription RPC: grpc: context canceled [code 1/Canceled]
W230117 06:23:01.205921 1244063 sql/sqlliveness/slinstance/slinstance.go:312  [T1,n1] 1837  exiting heartbeat loop
W230117 06:23:01.207933 1702396 sql/sqlliveness/slinstance/slinstance.go:312  [T1,nsql3] 1846  exiting heartbeat loop
W230117 06:23:01.207831 1744142 ccl/kvccl/kvtenantccl/connector.go:235  [T10,nsql4,tenant-connector] 1847  error consuming GossipSubscription RPC: grpc: context canceled [code 1/Canceled]
W230117 06:23:01.209606 1664405 ccl/kvccl/kvtenantccl/setting_overrides.go:46  [T10,nsql2,tenant-connector] 1848  error consuming TenantSettings RPC: grpc: context canceled [code 1/Canceled]
W230117 06:23:01.218821 1464510 sql/sqlliveness/slinstance/slinstance.go:312  [T1,n4] 1849  exiting heartbeat loop
W230117 06:23:01.209290 1665428 sql/sqlliveness/slinstance/slinstance.go:312  [T1,nsql2] 1851  exiting heartbeat loop
W230117 06:23:01.214064 1524877 jobs/registry.go:1007  [T10,nsql1] 1850  canceling all adopted jobs due to stopper quiescing
I230117 06:23:01.211875 1703210 sql/stats/automatic_stats.go:508  [T10,nsql3] 1852  quiescing auto stats refresher
W230117 06:23:01.212164 1850575 kv/kvserver/closedts/sidetransport/receiver.go:139  [T1,n3] 1853  closed timestamps side-transport connection dropped from node: 1
I230117 06:23:01.214009 1524694 sql/stats/automatic_stats.go:508  [T10,nsql1] 1854  quiescing auto stats refresher
W230117 06:23:01.212075 1425482 sql/sqlliveness/slinstance/slinstance.go:297  [T1,n3] 1855  exiting heartbeat loop
W230117 06:23:01.218654 1666080 jobs/registry.go:1007  [T10,nsql2] 1856  canceling all adopted jobs due to stopper quiescing
W230117 06:23:01.208323 1701419 ccl/kvccl/kvtenantccl/connector.go:235  [T10,nsql3,tenant-connector] 1857  error consuming GossipSubscription RPC: grpc: context canceled [code 1/Canceled]
W230117 06:23:01.221058 1385368 sql/sqlliveness/slinstance/slinstance.go:312  [T1,n2] 1858  exiting heartbeat loop
W230117 06:23:01.221275 1702396 sql/sqlliveness/slinstance/slinstance.go:297  [T1,nsql3] 1860  exiting heartbeat loop
W230117 06:23:01.221507 1464510 sql/sqlliveness/slinstance/slinstance.go:297  [T1,n4] 1861  exiting heartbeat loop
I230117 06:23:01.224650 1548833 jobs/registry.go:1427  [T10,nsql1] 1862  AUTO SPAN CONFIG RECONCILIATION job 831868689912561665: stepping through state succeeded
W230117 06:23:01.207513 1703883 jobs/registry.go:1007  [T10,nsql3] 1859  canceling all adopted jobs due to stopper quiescing
W230117 06:23:01.224455 1385368 sql/sqlliveness/slinstance/slinstance.go:297  [T1,n2] 1863  exiting heartbeat loop
W230117 06:23:01.221597 1520437 ccl/kvccl/kvtenantccl/connector.go:235  [T10,nsql1,tenant-connector] 1864  error consuming GossipSubscription RPC: grpc: context canceled [code 1/Canceled]
W230117 06:23:01.208458 1385725 jobs/registry.go:1007  [T1,n2] 1865  canceling all adopted jobs due to stopper quiescing
W230117 06:23:01.222181 1521511 sql/sqlliveness/slinstance/slinstance.go:312  [T1,nsql1] 1866  exiting heartbeat loop
W230117 06:23:01.225214 1521511 sql/sqlliveness/slinstance/slinstance.go:297  [T1,nsql1] 1867  exiting heartbeat loop
W230117 06:23:01.218879 1426578 jobs/registry.go:1007  [T1,n3] 1868  canceling all adopted jobs due to stopper quiescing
I230117 06:23:01.213628 1244130 sql/stats/automatic_stats.go:508  [T1,n1] 1869  quiescing auto stats refresher
W230117 06:23:01.212458 1845360 kv/kvserver/intentresolver/intent_resolver.go:799  [T1] 1870  failed to gc transaction record: could not GC completed transaction anchored at /Local/Range/Tenant/10/RangeDescriptor: node unavailable; try another peer
W230117 06:23:01.212632 1744764 sql/sqlliveness/slinstance/slinstance.go:297  [T1,nsql4] 1871  exiting heartbeat loop
W230117 06:23:01.208781 1744147 ccl/kvccl/kvtenantccl/setting_overrides.go:46  [T10,nsql4,tenant-connector] 1872  error consuming TenantSettings RPC: grpc: context canceled [code 1/Canceled]
I230117 06:23:01.212924 1271192 jobs/registry.go:1427  [T1,n1] 1873  AUTO SPAN CONFIG RECONCILIATION job 831868681440428033: stepping through state succeeded
W230117 06:23:01.213459 1244169 jobs/registry.go:1007  [T1,n1] 1874  canceling all adopted jobs due to stopper quiescing
I230117 06:23:01.218621 1665982 sql/stats/automatic_stats.go:508  [T10,nsql2] 1876  quiescing auto stats refresher
W230117 06:23:01.219487 1851244 kv/kvserver/closedts/sidetransport/receiver.go:139  [T1,n2] 1875  closed timestamps side-transport connection dropped from node: 1
W230117 06:23:01.208550 1701424 ccl/kvccl/kvtenantccl/setting_overrides.go:46  [T10,nsql3,tenant-connector] 1877  error consuming TenantSettings RPC: grpc: context canceled [code 1/Canceled]
W230117 06:23:01.222588 1244063 sql/sqlliveness/slinstance/slinstance.go:297  [T1,n1] 1878  exiting heartbeat loop
W230117 06:23:01.222840 1665428 sql/sqlliveness/slinstance/slinstance.go:297  [T1,nsql2] 1879  exiting heartbeat loop
W230117 06:23:01.222976 1520442 ccl/kvccl/kvtenantccl/setting_overrides.go:46  [T10,nsql1,tenant-connector] 1880  error consuming TenantSettings RPC: grpc: context canceled [code 1/Canceled]
W230117 06:23:01.227672 1548833 jobs/adopt.go:455  [T1] 1881  could not clear job claim: clear-job-claim: node unavailable; try another peer
W230117 06:23:01.237460 1271192 jobs/adopt.go:455  [T1,n1] 1882  could not clear job claim: clear-job-claim: failed to send RPC: sending to all replicas failed; last error: node unavailable; try another peer
    --- FAIL: TestSecondaryTenantFollowerReadsRouting/valid-latency-func=true (6.99s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #92379 ccl/kvccl/kvfollowerreadsccl: TestSecondaryTenantFollowerReadsRouting failed [C-test-failure O-robot T-kv branch-release-22.2]
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestSecondaryTenantFollowerReadsRouting.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-23463

--- a/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
+++ b/pkg/ccl/kvccl/kvfollowerreadsccl/followerreads_test.go
@@ -834,7 +834,6 @@ func TestFollowerReadsWithStaleDescriptor(t *testing.T) {
 // where it needs to be estimated using node localities.
 func TestSecondaryTenantFollowerReadsRouting(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 95338, "flaky test")
 	defer utilccl.TestingEnableEnterprise()()
 
 	skip.UnderStressRace(t, "times out")
@@ -973,6 +972,7 @@ func TestSecondaryTenantFollowerReadsRouting(t *testing.T) {
 
 		startKey := keys.MakeSQLCodec(serverutils.TestTenantID()).TenantPrefix()
 		tc.AddVotersOrFatal(t, startKey, tc.Target(1), tc.Target(2))
+		tc.WaitForVotersOrFatal(t, startKey, tc.Target(1), tc.Target(2))
 		desc := tc.LookupRangeOrFatal(t, startKey)
 		require.Equal(t, []roachpb.ReplicaDescriptor{
 			{NodeID: 1, StoreID: 1, ReplicaID: 1},

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -852,6 +852,14 @@ func (tc *TestCluster) WaitForVoters(
 	return tc.waitForNewReplicas(startKey, true /* waitForVoter */, targets...)
 }
 
+// WaitForVotersOrFatal is the same as WaitForVoters but it will Fatal the test
+// on error.
+func (tc *TestCluster) WaitForVotersOrFatal(
+	t testing.TB, startKey roachpb.Key, targets ...roachpb.ReplicationTarget,
+) {
+	require.NoError(t, tc.WaitForVoters(startKey, targets...))
+}
+
 // waitForNewReplicas waits for each of the targets to have a fully initialized
 // replica of the range indicated by startKey.
 //

